### PR TITLE
Book feature testing

### PR DESCRIPTION
**Pull request by abbymapes**
Added testing for the following:

BookAgentTest.java
- BookAgent intent handling from Dialogflow
- Tests whether fulfillment and display are correct for all Books intents
- Populates database with Indices, BookQuery, and stored Book information to ensure that memory handling is correct

BookQueryTest.java
- Tests whether creation of BookQuery objects are accurate given the parameters

BookCreationTest.java
- Tests whether calls to Google Books API match the specifications
- Tests whether the creation of Book objects is correct

Added helper code in classes to create Book objects for testing, pass SessionIDs as a part of the testing framework, and other handling for Book testing

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/agents/Memory.java [hunk 1]
```diff
@@ -29,7 +29,7 @@
   private DatastoreService datastore;
   private UserService userService;
   private String listName;
-  private ArrayList<String> items = new ArrayList<>();
+  private ArrayList<String> items = new ArrayList<>();;
```
Comment (aliu917) [inline portfolio/src/main/java/com/google/sps/agents/Memory.java:32]: Extra semicolon?

File: portfolio/src/main/java/com/google/sps/data/Book.java [hunk 2]
```diff
@@ -89,6 +89,30 @@ private Book(Volume volume) {
...
+   * @param description String description of book
+   * @param embeddable Bool indicating whether book is embeddable
+   * @param ISBN String unique ISBN number
+   */
+  public Book(String title, String authors, String description, Boolean embeddable, String ISBN) {
+    this.title = title;
+    this.authors = authors;
+    this.description = description;
+    this.embeddable = embeddable;
+    this.isbn = isbn;
```
Comment (aliu917) [inline portfolio/src/main/java/com/google/sps/data/Book.java:107]: Should this be this.isbn = ISBN because the parameter input is all caps?

Comment (abbymapes) [reply portfolio/src/main/java/com/google/sps/data/Book.java:107]: Fixed in next commit

File: portfolio/src/main/java/com/google/sps/agents/Memory.java (hunk 1 above)
Comment (abbymapes) [reply portfolio/src/main/java/com/google/sps/agents/Memory.java:32]: Fixed in next commit

File: portfolio/src/main/java/com/google/sps/data/Book.java [hunk 3]
```diff
@@ -89,6 +89,30 @@ private Book(Volume volume) {
...
+   * Public Book constructor, used for testing purposes. The constructor will set the following
+   * fields from the parameters. The rest of the properties will be set to an Empty String.
+   *
+   * @param title String title of book
+   * @param authors String authors of book
+   * @param description String description of book
+   * @param embeddable Bool indicating whether book is embeddable
+   * @param ISBN String unique ISBN number
+   */
+  public Book(String title, String authors, String description, Boolean embeddable, String ISBN) {
```
Comment (marcusschwartz) [inline portfolio/src/main/java/com/google/sps/data/Book.java:102]: The style guide seems to say this should be "Isbn" (https://google.github.io/styleguide/javaguide.html#s5.3-camel-case).  I'd argue that isbn is better than Isbn, personally.  In any case it seems like you should be consistent with class variable and constructor parameter names.

File: portfolio/src/main/java/com/google/sps/agents/BooksAgent.java [hunk 4]
```diff
@@ -269,12 +269,12 @@ private int getNextStartIndex(int prevIndex, int total) {
     return -1;
   }
 
-  private String bookToString(Book book) {
+  public static String bookToString(Book book) {
```
Comment (marcusschwartz) [inline portfolio/src/main/java/com/google/sps/agents/BooksAgent.java:272]: why not bookToJson?

File: portfolio/src/main/java/com/google/sps/agents/BooksAgent.java [hunk 5]
```diff
@@ -269,12 +269,12 @@ private int getNextStartIndex(int prevIndex, int total) {
...
     Gson gson = new Gson();
     return gson.toJson(book);
   }
 
-  private String bookListToString(ArrayList<Book> books) {
+  public static String bookListToString(ArrayList<Book> books) {
```
Comment (marcusschwartz) [inline portfolio/src/main/java/com/google/sps/agents/BooksAgent.java:277]: ditto, bookListToJson?

File: portfolio/src/main/java/com/google/sps/servlets/BookAgentServlet.java [hunk 6]
```diff
@@ -72,21 +73,23 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
    * @param queryID unique ID (within sessionID) for current query
    * @param parameterMap map containing parameters needed, if any, by BookAgent
    * @param languageCode language code
+   * @param datastore DataStore service to use
+   * @param userService UserService to use
```
Comment (marcusschwartz) [inline portfolio/src/main/java/com/google/sps/servlets/BookAgentServlet.java:77]: extra parameter, also the order in the comments slightly differs from the order in the params

File: portfolio/src/test/java/com/google/sps/TestHelper.java [hunk 7]
```diff
@@ -122,6 +142,28 @@ public Output getOutput() throws IOException {
     return output;
   }
 
+  /**
+   * Populates customizes datastore with desired Indices object.
```
Comment (marcusschwartz) [inline portfolio/src/test/java/com/google/sps/TestHelper.java:146]: this comment seems to be re-used a bunch.. is it accurate?

File: portfolio/src/main/java/com/google/sps/data/Book.java (hunk 3 above)
Comment (abbymapes) [reply portfolio/src/main/java/com/google/sps/data/Book.java:102]: Fixed in next commit

File: portfolio/src/main/java/com/google/sps/servlets/BookAgentServlet.java (hunk 6 above)
Comment (abbymapes) [reply portfolio/src/main/java/com/google/sps/servlets/BookAgentServlet.java:77]: Fixed in next commit

File: portfolio/src/test/java/com/google/sps/TestHelper.java (hunk 7 above)
Comment (abbymapes) [reply portfolio/src/test/java/com/google/sps/TestHelper.java:146]: Fixed in next commit

**Pull request merged**